### PR TITLE
L041: Fix bug when there is a newline after the select clause modifier

### PR DESCRIPTION
**Pull request by jmc-bbk**
### Issue

When there is a newline after the `SELECT CLAUSE MODIFIER`, rule L041 will require introduce unnecessary whitespace. This requires a second run of `sqlfluff fix <>` to resolve.

Example:

``` 
-- test.sql
CREATE OR REPLACE TABLE myschema.mytable AS (
  SELECT
  DISTINCT
    cola
    , colb
  FROM myschema.mytable
);
```

Using the CLI:

```
== [test.sql] FAIL                                                                                                                                                                                              
L:   2 | P:   3 | L041 | 'SELECT' modifiers (e.g. 'DISTINCT') must be on the same                                                                                                                               
                       | line as 'SELECT'.
==== fixing violations ====
1 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n] Y
```
Output:

```
CREATE OR REPLACE TABLE myschema.mytable AS (
  SELECT DISTINCT  
    cola
    , colb
  FROM myschema.mytable
);
```

Using the CLI again:

```
==== finding fixable violations ====
== [test.sql] FAIL                                                                                                                                                                                              
L:   2 | P:  18 | L001 | Unnecessary trailing whitespace.                                                                                                                                                       
==== fixing violations ====
1 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n] 
```

### Brief summary of the changes made

- Added a variable that has any leading whitespace between the `SELECT` and the `SELECT CLAUSE MODIFIER`.
- Added a conditional so that if there is a `NEWLINE` after the `SELECT CLAUSE MODIFIER` we delete this leading whitespace in addition to (the already deleted) newlines.

### Are there any other side effects of this change that we should be aware of?

- None that I am aware of (first contribution)!

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/rules/L041.py [hunk 1]
```diff
@@ -93,9 +101,24 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
...
+            fixes.extend((LintFix.delete(s) for s in leading_newline_segments))
+        # If there is a newline after the select clause modifier then delete both the
+        # newlines and whitespace between the select keyword and select clause modifier.
+        else:
+            fixes.extend(
+                (
+                    LintFix.delete(s)
+                    for s in leading_newline_segments + leading_whitespace_segments
+                )
+            )
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L041.py:120]: Does this need to be two different statements? Will `leading_whitespace_segments` not just be blank? So any harm in always using the section option?

Comment (jmc-bbk) [reply src/sqlfluff/rules/L041.py:120]: Hey @tunetheweb, I initially tried that and found it broke test 1 & 2. It was incorrectly removing the white space from the trailing `SELECT ELEMENT`.

I’m not at my computer at the moment; but I can provide an example tomorrow!

Comment (tunetheweb) [reply src/sqlfluff/rules/L041.py:120]: OK let's leave as is then.

File: src/sqlfluff/rules/L041.py [hunk 2]
```diff
@@ -93,9 +101,24 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
...
-        # and delete the original select clause modifier.
-        fixes.extend((LintFix.delete(s) for s in leading_newline_segments))
+
+        # Delete original newlines and whitespace between select keyword
+        # and select clause modifier.
+
+        # If there is not a newline after the select clause modifier then delete
+        # newlines between the select keyword and select clause modifier.
+        if not trailing_newline_segments:
+            fixes.extend((LintFix.delete(s) for s in leading_newline_segments))
```
Comment (barrywhart) [inline src/sqlfluff/rules/L041.py:111]: Nitpick: IIUC, the parentheses are not needed here. `LintFix.delete(s) for s in leading_newline_segments` is a generator expression, which can be passed directly to `extend()`.
```suggestion
            fixes.extend(LintFix.delete(s) for s in leading_newline_segments)
```

File: src/sqlfluff/rules/L041.py [hunk 3]
```diff
@@ -93,9 +101,24 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
...
+        # If there is a newline after the select clause modifier then delete both the
+        # newlines and whitespace between the select keyword and select clause modifier.
+        else:
+            fixes.extend(
+                (
```
Comment (barrywhart) [inline src/sqlfluff/rules/L041.py:116]: Ditto the above comment.

**Review by barrywhart — APPROVED** [review]
Looks good. A couple of nitpicky suggestions for change (not 100% sure I'm correct about this, but I think so).

File: src/sqlfluff/rules/L041.py (hunk 2 above)
Comment (tunetheweb) [reply src/sqlfluff/rules/L041.py:111]: Test this and you are correct. Will change and merge.

**Pull request merged**